File: fixtures/projects/e2e-app/src/workers/notification.ts
import type { Logger } from "../utils/logger";
import { sendEmail } from "./email";

export type NotificationPayload = {
	channel: "email" | "sms" | "push";
	recipient: string;
	message: string;
};

export type NotificationResult = {
	notificationId: string;
	delivered: boolean;
};

let notificationCounter = 0;

export async function enqueueNotification(payload: NotificationPayload, logger: Logger): Promise<NotificationResult> {
	const notificationId = `notif-${++notificationCounter}`;
	logger.info("Enqueuing notification", { notificationId, channel: payload.channel });

	if (payload.channel === "email") {
		await sendEmail({ to: payload.recipient, subject: "Notification", body: payload.message, priority: "normal" }, logger);
	}

	return { notificationId, delivered: true };
}

export function formatNotification(result: NotificationResult): string {
	return `Notification ${result.notificationId}: ${result.delivered ? "delivered" : "pending"}`;
}

File: fixtures/projects/e2e-python/src/workers/notification.py
from src.logging.logger import AppLogger
from src.workers.email import send_email


def enqueue_notification(
    payload: dict[str, str], logger: AppLogger
) -> dict[str, str | bool]:
    logger.info("queueing notification", channel=payload["channel"])
    if payload["channel"] == "email":
        send_email(
            {
                "to": payload["recipient"],
                "subject": "notification",
                "body": payload["message"],
                "priority": "normal",
            },
            logger,
        )
    return {"notification_id": f"notif-{payload['channel']}", "queued": True}


def format_notification(result: dict[str, str | bool]) -> str:
    state = "queued" if result.get("queued") else "failed"
    return f"notification {result['notification_id']} is {state}"
